File: sdks/go/providers/minimax/socket.go
package minimax

import (
	"context"
	"encoding/json"
	"errors"
	schema "github.com/speechswitch/client/sdks/go/generated/minimax"
	out "github.com/speechswitch/client/sdks/go/generated/minimax_output"
	"github.com/speechswitch/client/sdks/go/runtime"
	"io"
	"strings"
	"sync"
	"sync/atomic"
	"time"
	"unicode/utf16"
)

type socketResult struct {
	packet packet
	err    error
}
type socketStream struct {
	parent, ctx                        context.Context
	cancel                             context.CancelFunc
	stop                               func() bool
	socket                             runtime.WebSocketLike
	config                             configuration
	validate                           runtime.InputValidator
	limit                              int
	session                            string
	connection                         runtime.Optional[string]
	once                               sync.Once
	mutex, lifecycle                   sync.Mutex
	closeError                         error
	started, done, finishing, clearing atomic.Bool
	producerIdle                       <-chan struct{}
	handshakeResult                    <-chan socketResult
	producerResult                     <-chan error
	pending                            <-chan socketResult
	resume                             chan struct{}
	initialized, ready, terminal       bool
}

func (s *socketStream) closeResources() error {
	s.once.Do(func() {
		s.lifecycle.Lock()
		s.cancel()
		idle := s.producerIdle
		s.lifecycle.Unlock()
		if s.started.Load() && !s.done.Load() {
			// Cancellation must not wait indefinitely for a blocked writer.
			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Millisecond)
			sent := make(chan struct{})
			go func() { _ = s.send(ctx, map[string]any{"event": "task_cancel"}); close(sent) }()
			select {
			case <-sent:
			case <-ctx.Done():
			}
			cancel()
		}
		s.closeError = s.socket.Close()
		go func() { <-idle; _ = s.config.input.Close() }()
	})
	return s.closeError
}
func (s *socketStream) Close() error { s.stop(); return s.closeResources() }
func (s *socketStream) send(ctx context.Context, value map[string]any) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	data, err := json.Marshal(value)
	if err != nil {
		return err
	}
	return s.socket.Send(ctx, runtime.WebSocketText(data))
}
func (s *socketStream) receive() (packet, error) {
	message, err := s.socket.Receive(s.ctx)
	if err == io.EOF {
		return packet{}, errors.New("MiniMax WebSocket closed before task_finished")
	}
	if err != nil {
		return packet{}, err
	}
	text, ok := message.(runtime.WebSocketText)
	if !ok {
		return packet{}, errors.New("MiniMax WebSocket messages must be JSON text")
	}
	if len(text) > s.limit {
		return packet{}, errors.New("MiniMax message exceeds MaxMessageBytes")
	}
	p, err := decodePacket([]byte(text))
	if err != nil {
		return p, err
	}
	if p.code != 0 {
		return p, &Error{Message: p.message, Code: runtime.Some(p.code)}
	}
	return p, nil
}
func (s *socketStream) handshake() (packet, error) {
	connected, err := s.receive()
	if err != nil {
		return packet{}, err
	}
	if !connected.event.Present || connected.event.Value != "connected_success" {
		return packet{}, errors.New("MiniMax did not acknowledge the connection")
	}
	config := map[string]any{}
	for k, v := range s.config.wire {
		config[k] = v
	}
	config["event"], config["session_id"], config["subtitle_enable"] = "task_start", s.session, false
	s.started.Store(true)
	if err = s.send(s.ctx, config); err != nil {
		return packet{}, err
	}
	started, err := s.receive()
	if err != nil {
		return packet{}, err
	}
	if !started.event.Present || started.event.Value != "task_started" {
		return packet{}, errors.New("MiniMax did not acknowledge task_start")
	}
	if started.session.Present && started.session.Value != s.session {
		return packet{}, errors.New("MiniMax returned an unexpected session ID")
	}
	connection := connected.connection
	if !connection.Present {
		connection = connected.session
	}
	if connection.Present && started.connection.Present && connection.Value != started.connection.Value {
		return packet{}, errors.New("MiniMax returned an unexpected connection ID")
	}
	return packet{connection: connection}, nil
}
func (s *socketStream) start() {
	s.lifecycle.Lock()
	defer s.lifecycle.Unlock()
	if s.ctx.Err() != nil || s.initialized {
		return
	}
	s.initialized = true
	idle, result := make(chan struct{}), make(chan socketResult, 1)
	s.producerIdle, s.handshakeResult = idle, result
	go func() { p, err := s.handshake(); close(idle); result <- socketResult{p, err} }()
}

// Match ECMAScript trim and UTF-16 limits so token boundaries survive across SDKs.
const whitespace = " \t\n\r\v\f\u00a0\u1680\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u2028\u2029\u202f\u205f\u3000\ufeff"

func (s *socketStream) produce() error {
	pendingSpace := ""
	awaitingCancel := false
	for {
		if err := s.ctx.Err(); err != nil {
			return err
		}
		item, err := s.config.input.Next(s.ctx)
		if cause := s.ctx.Err(); cause != nil {
			return cause
		}
		if err != nil && err != io.EOF {
			return err
		}
		if err == nil {
			if err = s.validate(item); err != nil {
				return err
			}
		}
		if awaitingCancel {
			select {
			case <-s.resume:
			case <-s.ctx.Done():
				return s.ctx.Err()
			}
			awaitingCancel = false
		}
		if err == io.EOF {
			s.finishing.Store(true)
			return s.send(s.ctx, map[string]any{"event": "task_finish"})
		}
		switch v := item.(type) {
		case *schema.TtsRequestStreamingText12421ea0TextItemAsString:
			item = *v
		case *schema.TtsRequestStreamingText12421ea0TextItemAsClear:
			item = *v
		case *schema.TtsRequestStreamingText12421ea0TextItemAsFlush:
			item = *v
		}
		switch v := item.(type) {
		case schema.TtsRequestStreamingText12421ea0TextItemAsString:
			if len(utf16.Encode([]rune(v.Value))) >= 10000 {
				return errors.New("MiniMax text pieces must contain fewer than 10000 UTF-16 code units")
			}
			text := pendingSpace + v.Value
			pendingSpace = ""
			if strings.Trim(text, whitespace) == "" {
				if len(utf16.Encode([]rune(text))) >= 10000 {
					return errors.New("MiniMax pending whitespace exceeds a native message")
				}
				pendingSpace = text
			} else {
				for text != "" {
					end, units := 0, 0
					for i, r := range text {
						width := 1
						if r > 0xffff {
							width = 2
						}
						if units+width > 9999 {
							break
						}
						units += width
						end = i + len(string(r))
					}
					piece := text[:end]
					text = text[end:]
					if strings.Trim(piece, whitespace) != "" {
						if err = s.send(s.ctx, map[string]any{"event": "task_continue", "text": piece}); err != nil {
							return err
						}
					} else if text != "" {
						return errors.New("MiniMax pending whitespace exceeds a native message")
					} else {
						pendingSpace = piece
					}
				}
			}
		case schema.TtsRequestStreamingText12421ea0TextItemAsClear:
			pendingSpace = ""
			s.clearing.Store(true)
			awaitingCancel = true
			if err = s.send(s.ctx, map[string]any{"event": "task_cancel"}); err != nil {
				return err
			}
		case schema.TtsRequestStreamingText12421ea0TextItemAsFlush:
			pendingSpace = ""
			if err = s.send(s.ctx, map[string]any{"event": "task_flush"}); err != nil {
				return err
			}
		}
	}
}
func (s *socketStream) startProducer() {
	s.lifecycle.Lock()
	defer s.lifecycle.Unlock()
	if s.ctx.Err() != nil {
		return
	}
	idle, result := make(chan struct{}), make(chan error, 1)
	s.producerIdle, s.producerResult = idle, result
	go func() { err := s.produce(); close(idle); result <- err }()
}
func (s *socketStream) Next(ctx context.Context) (out.SynthesisItem, error) {
	stop := context.AfterFunc(ctx, func() { s.closeResources() })
	defer stop()
	s.mutex.Lock()
	defer s.mutex.Unlock()
	if s.terminal || s.done.Load() {
		return nil, io.EOF
	}
	item, err := s.next(ctx)
	if cause := ctx.Err(); cause != nil {
		item, err = nil, cause
	} else if cause := s.parent.Err(); cause != nil {
		item, err = nil, cause
	} else if s.ctx.Err() != nil && !s.done.Load() {
		item, err = nil, io.EOF
	}
	if err != nil {
		s.terminal = true
		s.Close()
	}
	return item, err
}
func (s *socketStream) next(ctx context.Context) (out.SynthesisItem, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	if s.ctx.Err() != nil {
		return nil, io.EOF
	}
	s.start()
	if !s.ready {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-s.ctx.Done():
			return nil, io.EOF
		case result := <-s.handshakeResult:
			if result.err != nil {
				return nil, result.err
			}
			s.connection = result.packet.connection
			s.ready = true
			s.startProducer()
		}
	}
	for {
		if s.pending == nil {
			result := make(chan socketResult, 1)
			s.pending = result
			go func() { p, err := s.receive(); result <- socketResult{p, err} }()
		}
		var result socketResult
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-s.ctx.Done():
			return nil, io.EOF
		case err := <-s.producerResult:
			s.producerResult = nil
			if err != nil {
				return nil, err
			}
			continue
		case result = <-s.pending:
			s.pending = nil
		}
		if result.err != nil {
			return nil, result.err
		}
		p := result.packet
		if p.session.Present && p.session.Value != s.session {
			return nil, errors.New("MiniMax returned an unexpected session ID")
		}
		if s.connection.Present && p.connection.Present && p.connection.Value != s.connection.Value {
			return nil, errors.New("MiniMax returned an unexpected connection ID")
		}
		event := p.event.Value
		if !p.event.Present && (p.audioPresent || p.final.Present) {
			event = "task_continued"
		}
		if event == "task_failed" {
			message := p.message
			if message == "" {
				message = "MiniMax synthesis task failed"
			}
			return nil, &Error{Message: message, Code: runtime.Some(p.code)}
		}
		if event == "task_finished" && s.clearing.Load() {
			return nil, errors.New("MiniMax ended the session before acknowledging cancellation")
		}
		if event == "task_canceled" {
			if !s.clearing.Swap(false) {
				return nil, errors.New("MiniMax returned an unsolicited cancel acknowledgement")
			}
			s.resume <- struct{}{}
			return out.SynthesisItemAsClear{}, nil
		}
		if s.clearing.Load() {
			continue
		}
		switch event {
		case "task_finished":
			if !s.finishing.Load() {
				return nil, errors.New("MiniMax ended the session before task_finish")
			}
			if s.producerResult != nil {
				select {
				case <-ctx.Done():
					return nil, ctx.Err()
				case <-s.ctx.Done():
					return nil, io.EOF
				case err := <-s.producerResult:
					s.producerResult = nil
					if err != nil {
						return nil, err
					}
				}
			}
			s.done.Store(true)
			s.Close()
			return out.SynthesisItemAsDone{Value: out.MiniMaxDoneEvent{TraceId: p.trace, Usage: p.usage}}, nil
		case "task_flushed":
			id := s.session
			if p.trace.Present {
				id = p.trace.Value
			}
			return out.SynthesisItemAsFlush{Value: out.FlushEvent{CorrelationId: id, InputGroupId: runtime.Some(s.session)}}, nil
		case "sentence_start", "sentence_end", "task_continued":
			envelope := out.MiniMaxEnvelope{Correlation: out.MiniMaxEnvelopeCorrelationAsOrdered{}, InputGroupId: runtime.Some(s.session), CorrelationId: p.trace, TraceId: p.trace, Timestamps: []out.MiniMaxTimestamp{}}
			if event == "sentence_start" {
				var boundary out.MiniMaxEnvelopeSentenceBoundary = out.MiniMaxEnvelopeSentenceBoundaryAsStart{}
				envelope.SentenceBoundary = runtime.Some(boundary)
			} else if event == "sentence_end" {
				var boundary out.MiniMaxEnvelopeSentenceBoundary = out.MiniMaxEnvelopeSentenceBoundaryAsEnd{}
				envelope.SentenceBoundary = runtime.Some(boundary)
			} else {
				if len(p.audio) == 0 && !p.final.Value && !p.usage.Present {
					continue
				}
				if len(p.audio) > 0 {
					envelope.Audio = runtime.Some(p.audio)
				}
				if p.final.Value {
					var complete out.MiniMaxEnvelopeRequestComplete = out.MiniMaxEnvelopeRequestCompleteAsTrue{}
					envelope.RequestComplete = runtime.Some(complete)
				}
				envelope.Usage = p.usage
			}
			return out.SynthesisItemAsOrderedOrTimeline{Value: envelope}, nil
		}
	}
}

File: sdks/rust/src/providers/minimax/socket.rs
use super::{
    protocol::{self, failure, Error, Packet},
    settings::{Map, Settings, Validator},
    Input, SynthesisItem,
};
use crate::{
    generated::minimax_output::*,
    http::TransportError,
    json,
    runtime::{JsonValue, StreamingInput},
    websocket::{Message, Socket},
};
use std::{
    collections::VecDeque,
    sync::Arc,
    task::{Context, Poll, Wake, Waker},
};

pub(super) struct SocketStream {
    // Drop the socket before producer cleanup.
    socket: Socket,
    input: StreamingInput<Input>,
    validate: Validator,
    create: Option<Map>,
    session: String,
    connection: Option<String>,
    limit: usize,
    connected: bool,
    started: bool,
    acknowledged: bool,
    writing: bool,
    finishing: bool,
    clearing: bool,
    queued: Option<Option<Input>>,
    pieces: VecDeque<String>,
    whitespace: String,
    final_packet: Option<Packet>,
    prefer_output: bool,
    pub done: bool,
}
enum Step {
    Progress,
    Item(SynthesisItem),
}
impl SocketStream {
    pub fn new(
        socket: Socket,
        mut settings: Settings,
        session: String,
        limit: usize,
        validate: Validator,
    ) -> Self {
        settings
            .wire
            .insert("event".into(), JsonValue::String("task_start".into()));
        settings
            .wire
            .insert("session_id".into(), JsonValue::String(session.clone()));
        settings
            .wire
            .insert("subtitle_enable".into(), JsonValue::Bool(false));
        Self {
            socket,
            input: settings.input.take().unwrap(),
            validate,
            create: Some(settings.wire),
            session,
            connection: None,
            limit,
            connected: false,
            started: false,
            acknowledged: false,
            writing: false,
            finishing: false,
            clearing: false,
            queued: None,
            pieces: VecDeque::new(),
            whitespace: String::new(),
            final_packet: None,
            prefer_output: true,
            done: false,
        }
    }
    fn send(&mut self, fields: Map) -> Result<(), TransportError> {
        let mut data = String::new();
        json::write(&JsonValue::Object(fields), &mut data)?;
        self.socket.as_mut().start_send(Message::Text(data))?;
        self.writing = true;
        Ok(())
    }
    pub fn next(
        &mut self,
        cx: &mut Context<'_>,
    ) -> Poll<Result<Option<SynthesisItem>, TransportError>> {
        if self.final_packet.is_some() {
            if self.writing {
                match self.socket.as_mut().poll_flush(cx) {
                    Poll::Pending => return Poll::Pending,
                    Poll::Ready(v) => v?,
                }
                self.writing = false;
            }
            let p = self.final_packet.take().unwrap();
            self.done = true;
            return Poll::Ready(Ok(Some(SynthesisItem::Done(MiniMaxDoneEvent {
                event: Default::default(),
                trace_id: p.trace,
                usage: p.usage,
            }))));
        }
        for output in if self.prefer_output {
            [true, false]
        } else {
            [false, true]
        } {
            match if output {
                self.receive(cx)
            } else {
                self.produce(cx)
            } {
                Poll::Pending => {}
                Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
                Poll::Ready(Ok(step)) => {
                    self.prefer_output = !output;
                    if let Step::Item(item) = step {
                        return Poll::Ready(Ok(Some(item)));
                    }
                    cx.waker().wake_by_ref();
                    return Poll::Pending;
                }
            }
        }
        Poll::Pending
    }
    fn produce(&mut self, cx: &mut Context<'_>) -> Poll<Result<Step, TransportError>> {
        if self.writing {
            match self.socket.as_mut().poll_flush(cx) {
                Poll::Pending => return Poll::Pending,
                Poll::Ready(v) => v?,
            }
            self.writing = false;
            return Poll::Ready(Ok(Step::Progress));
        }
        if !self.connected {
            return Poll::Pending;
        }
        if let Some(create) = self.create.take() {
            self.started = true;
            self.send(create)?;
            return Poll::Ready(Ok(Step::Progress));
        }
        if !self.acknowledged || self.finishing {
            return Poll::Pending;
        }
        if let Some(piece) = self.pieces.pop_front() {
            self.send(Map::from([
                ("event".into(), JsonValue::String("task_continue".into())),
                ("text".into(), JsonValue::String(piece)),
            ]))?;
            return Poll::Ready(Ok(Step::Progress));
        }
        if self.queued.is_none() {
            match self.input.as_mut().poll_next(cx) {
                Poll::Pending => return Poll::Pending,
                Poll::Ready(Some(Err(e))) => return Poll::Ready(Err(e)),
                Poll::Ready(item) => {
                    let item = item.transpose()?;
                    if let Some(item) = &item {
                        (self.validate)(item, Some("text"))?;
                    }
                    self.queued = Some(item);
                }
            }
        }
        // One prefetched input lets a producer error interrupt a stalled cancel acknowledgement.
        if self.clearing {
            return Poll::Pending;
        }
        let Some(item) = self.queued.take().unwrap() else {
            self.finishing = true;
            self.send(Map::from([(
                "event".into(),
                JsonValue::String("task_finish".into()),
            )]))?;
            return Poll::Ready(Ok(Step::Progress));
        };
        match item {
            Input::String(value) => {
                if value.encode_utf16().count() >= 10000 {
                    return Poll::Ready(Err(failure(
                        "MiniMax text pieces must contain fewer than 10000 UTF-16 code units",
                    )));
                }
                let mut text = std::mem::take(&mut self.whitespace);
                text.push_str(&value);
                if text.trim_matches(WHITESPACE).is_empty() {
                    if text.encode_utf16().count() >= 10000 {
                        return Poll::Ready(Err(failure(
                            "MiniMax pending whitespace exceeds a native message",
                        )));
                    }
                    self.whitespace = text;
                } else {
                    let mut rest = text.as_str();
                    while !rest.is_empty() {
                        let mut end = 0;
                        let mut units = 0;
                        for (i, c) in rest.char_indices() {
                            if units + c.len_utf16() > 9999 {
                                break;
                            }
                            units += c.len_utf16();
                            end = i + c.len_utf8();
                        }
                        let (piece, tail) = rest.split_at(end);
                        if piece.trim_matches(WHITESPACE).is_empty() {
                            if !tail.is_empty() {
                                return Poll::Ready(Err(failure(
                                    "MiniMax pending whitespace exceeds a native message",
                                )));
                            }
                            self.whitespace = piece.into();
                        } else {
                            self.pieces.push_back(piece.into());
                        }
                        rest = tail;
                    }
                }
            }
            Input::Clear(_) => {
                self.whitespace.clear();
                self.clearing = true;
                self.send(Map::from([(
                    "event".into(),
                    JsonValue::String("task_cancel".into()),
                )]))?;
            }
            Input::Flush(_) => {
                self.whitespace.clear();
                self.send(Map::from([(
                    "event".into(),
                    JsonValue::String("task_flush".into()),
                )]))?;
            }
        }
        Poll::Ready(Ok(Step::Progress))
    }
    fn receive(&mut self, cx: &mut Context<'_>) -> Poll<Result<Step, TransportError>> {
        let data = match self.socket.as_mut().poll_receive(cx) {
            Poll::Pending => return Poll::Pending,
            Poll::Ready(None) => {
                return Poll::Ready(Err(failure(
                    "MiniMax WebSocket closed before task_finished",
                )))
            }
            Poll::Ready(Some(Err(e))) => return Poll::Ready(Err(e)),
            Poll::Ready(Some(Ok(Message::Binary(_)))) => {
                return Poll::Ready(Err(failure("MiniMax WebSocket messages must be JSON text")))
            }
            Poll::Ready(Some(Ok(Message::Text(v)))) => v,
        };
        if data.len() > self.limit {
            return Poll::Ready(Err(failure("MiniMax message exceeds max_message_bytes")));
        }
        let p = protocol::packet(&data)?;
        if p.code.is_some() {
            return Poll::Ready(Err(Box::new(Error {
                message: p.message,
                code: p.code,
                status: None,
                retry_after: None,
            })));
        }
        if !self.connected {
            if p.event.as_deref() != Some("connected_success") {
                return Poll::Ready(Err(failure("MiniMax did not acknowledge the connection")));
            }
            self.connection = p.connection.or(p.session);
            self.connected = true;
            return Poll::Ready(Ok(Step::Progress));
        }
        if p.session.as_ref().is_some_and(|v| v != &self.session) {
            return Poll::Ready(Err(failure("MiniMax returned an unexpected session ID")));
        }
        if p.connection
            .as_ref()
            .zip(self.connection.as_ref())
            .is_some_and(|(a, b)| a != b)
        {
            return Poll::Ready(Err(failure("MiniMax returned an unexpected connection ID")));
        }
        if !self.acknowledged {
            if !self.started || p.event.as_deref() != Some("task_started") {
                return Poll::Ready(Err(failure("MiniMax did not acknowledge task_start")));
            }
            self.acknowledged = true;
            return Poll::Ready(Ok(Step::Progress));
        }
        let event = p
            .event
            .as_deref()
            .unwrap_or(if p.audio.is_some() || p.final_audio.is_some() {
                "task_continued"
            } else {
                ""
            });
        if event == "task_failed" {
            return Poll::Ready(Err(Box::new(Error {
                message: if p.message.is_empty() {
                    "MiniMax synthesis task failed".into()
                } else {
                    p.message
                },
                code: p.code,
                status: None,
                retry_after: None,
            })));
        }
        if event == "task_finished" && self.clearing {
            return Poll::Ready(Err(failure(
                "MiniMax ended the session before acknowledging cancellation",
            )));
        }
        if event == "task_canceled" {
            if !self.clearing {
                return Poll::Ready(Err(failure(
                    "MiniMax returned an unsolicited cancel acknowledgement",
                )));
            }
            self.clearing = false;
            return Poll::Ready(Ok(Step::Item(SynthesisItem::Clear(ClearEvent {
                event: Default::default(),
            }))));
        }
        if self.clearing {
            return Poll::Ready(Ok(Step::Progress));
        }
        match event {
            "task_finished" => {
                if !self.finishing {
                    return Poll::Ready(Err(failure(
                        "MiniMax ended the session before task_finish",
                    )));
                }
                self.final_packet = Some(p);
                Poll::Ready(Ok(Step::Progress))
            }
            "task_flushed" => Poll::Ready(Ok(Step::Item(SynthesisItem::Flush(FlushEvent {
                event: Default::default(),
                correlation_id: p.trace.unwrap_or_else(|| self.session.clone()),
                input_group_id: Some(self.session.clone()),
            })))),
            "task_continued" | "sentence_start" | "sentence_end" => {
                let mut e = protocol::envelope(false, p.trace.clone());
                e.input_group_id = Some(self.session.clone());
                e.correlation_id = p.trace;
                match event {
                    "sentence_start" => {
                        e.sentence_boundary =
                            Some(MiniMaxEnvelopeSentenceBoundary::Start(Default::default()))
                    }
                    "sentence_end" => {
                        e.sentence_boundary =
                            Some(MiniMaxEnvelopeSentenceBoundary::End(Default::default()))
                    }
                    _ => {
                        e.audio = p.audio.filter(|v| !v.is_empty());
                        if p.final_audio == Some(true) {
                            e.request_complete =
                                Some(MiniMaxEnvelopeRequestComplete::True(Default::default()));
                        }
                        e.usage = p.usage;
                        if e.audio.is_none() && e.request_complete.is_none() && e.usage.is_none() {
                            return Poll::Ready(Ok(Step::Progress));
                        }
                    }
                }
                Poll::Ready(Ok(Step::Item(SynthesisItem::OrderedOrTimeline(e))))
            }
            _ => Poll::Ready(Ok(Step::Progress)),
        }
    }
}
// Match ECMAScript trim rather than Rust's broader Unicode whitespace definition.
const WHITESPACE: &[char] = &[
    ' ', '\t', '\n', '\r', '\u{b}', '\u{c}', '\u{a0}', '\u{1680}', '\u{2000}', '\u{2001}',
    '\u{2002}', '\u{2003}', '\u{2004}', '\u{2005}', '\u{2006}', '\u{2007}', '\u{2008}', '\u{2009}',
    '\u{200a}', '\u{2028}', '\u{2029}', '\u{202f}', '\u{205f}', '\u{3000}', '\u{feff}',
];
impl Drop for SocketStream {
    fn drop(&mut self) {
        // Never overlap an outstanding write, block Drop, or claim a delivered cancellation.
        if self.started && !self.done && !self.writing {
            if self
                .socket
                .as_mut()
                .start_send(Message::Text("{\"event\":\"task_cancel\"}".into()))
                .is_ok()
            {
                struct Noop;
                impl Wake for Noop {
                    fn wake(self: Arc<Self>) {}
                }
                let waker = Waker::from(Arc::new(Noop));
                let _ = self
                    .socket
                    .as_mut()
                    .poll_flush(&mut Context::from_waker(&waker));
            }
        }
    }
}
